support PyArrow timestamptz with Etc/UTC

syun64 opened the pull request:
This ended up being a lot simpler than I had thought it would be.

"Etc/UTC" is just another representation of UTC, and should be safe to support. This will allow us to support arrow dataframes produced with datetimes from libraries like pendulum that use this timezone.

Fixes: https://github.com/apache/iceberg-python/issues/863

HonahX commented on pyiceberg/io/pyarrow.py:
```diff
@@ -937,7 +937,7 @@ def primitive(self, primitive: pa.DataType) -> PrimitiveType:
                 raise TypeError(f"Unsupported precision for timestamp type: {primitive.unit}")
 
-            if primitive.tz == "UTC" or primitive.tz == "+00:00":
+            if primitive.tz in ("UTC", "+00:00", "Etc/UTC"):
```
In the future, shall we consider add other representations of UTC, such as `"Z"`?

Fokko replied:
Supernit:
```suggestion
            if primitive.tz in {"UTC", "+00:00", "Etc/UTC"}:
```

HonahX reviewed (approved):
LGTM! @syun64 Thanks for the quick fix!

syun64 commented on pyiceberg/table/__init__.py:
```diff
@@ -528,10 +528,6 @@ def append(self, df: pa.Table, snapshot_properties: Dict[str, str] = EMPTY_DICT)
-        # cast if the two schemas are compatible but not equal
-        table_arrow_schema = self._table.schema().as_arrow()
-        if table_arrow_schema != df.schema:
-            df = df.cast(table_arrow_schema)
```
Removed this cast function - `to_requested_schema` should be responsible for casting the types to their desired schema, instead of casting it here

syun64 commented:
@Fokko @HonahX - thank you for your reviews. I've updated the integration test to make the [check more comprehensive](https://github.com/apache/iceberg-python/pull/910/files#diff-7f3dd1244d08ce27c003cd091da10aa049f7bb0c7d5397acb4ec69767036accdR1063-R1074), and removed the [pyarrow table casting](https://github.com/apache/iceberg-python/pull/910/files#diff-23e8153e0fd497a9212215bd2067068f3b56fa071770c7ef326db3d3d03cee9bL590-L593) in `append` and `overwrite` so that each column can be cast with more targeted `safe` flag for each field.

syun64 commented on pyiceberg/io/pyarrow.py:
```diff
@@ -1320,7 +1321,16 @@ def _cast_if_needed(self, field: NestedField, values: pa.Array) -> pa.Array:
                     and values.type.unit == "ns"
                 ):
-                    return values.cast(target_type, safe=False)
+                    if target_type.tz == "UTC" and values.type.tz in UTC_ALIASES or not target_type.tz and not values.type.tz:
+                        return values.cast(target_type, safe=False)
+                if (
+                    pa.types.is_timestamp(target_type)
+                    and target_type.unit == "us"
+                    and pa.types.is_timestamp(values.type)
+                    and values.type.unit in {"s", "ms", "us"}
+                ):
+                    if target_type.tz == "UTC" and values.type.tz in UTC_ALIASES or not target_type.tz and not values.type.tz:
+                        return values.cast(target_type)
```
Does this look correct that we are casting the types in order to follow the Iceberg Spec for Parquet Physical and Logical Types? https://iceberg.apache.org/spec/#parquet

Fokko replied:
Yes, it looks good. Arrow writes `INT64` by default for timestamps: https://arrow.apache.org/docs/cpp/parquet.html#logical-types

Fokko commented on pyiceberg/io/pyarrow.py:
```diff
@@ -1320,7 +1321,16 @@ def _cast_if_needed(self, field: NestedField, values: pa.Array) -> pa.Array:
+                    and pa.types.is_timestamp(values.type)
+                    and values.type.unit in {"s", "ms", "us"}
+                ):
+                    if target_type.tz == "UTC" and values.type.tz in UTC_ALIASES or not target_type.tz and not values.type.tz:
```
Can we add parentheses?

syun64 replied:
Added

Fokko commented on pyiceberg/io/pyarrow.py:
```diff
@@ -1320,7 +1321,16 @@ def _cast_if_needed(self, field: NestedField, values: pa.Array) -> pa.Array:
+                    pa.types.is_timestamp(target_type)
+                    and target_type.unit == "us"
+                    and pa.types.is_timestamp(values.type)
+                    and values.type.unit in {"s", "ms", "us"}
```
When we're requesting `us`and the file provide a `us`, I don't think we need to cast?
```suggestion
                    and values.type.unit in {"s", "ms"}
```

syun64 replied:
Sorry this is a bit confusing, I agree 😓 . I included it here because I wanted to support casting `pa.timestamp('us', tz='Etc/UTC')` to `pa.timestamp('us', tz='UTC')` within the same condition.

I think we won't hit this condition if both the input and requested types are `pa.timestamp('us')` because we enter this block only if target_type and values.type are not equal:
https://github.com/apache/iceberg-python/blob/main/pyiceberg/io/pyarrow.py#L1315

Fokko replied:
I think we might need to do some more work here. In Iceberg we're rather strict on the distinction between Timestamp and TimestampTZ. A good way of showing this can be found here:

https://github.com/apache/iceberg-python/blob/aceed2ad9898eaa8b0cdddd883a195e58af5d0e3/pyiceberg/expressions/literals.py#L566-L572

This is when we parse a string from a literal, which often comes an expression: `dt >= '1925-05-22T00:00:00'`. If the value has a timezone, even UTC, we reject it as `Timestamp`. When it has a timestamp, we [normalize it to UTC and then store the integer](https://github.com/apache/iceberg-python/blob/aceed2ad9898eaa8b0cdddd883a195e58af5d0e3/pyiceberg/utils/datetime.py#L75).

syun64 replied:
Sounds good @Fokko- thank you for the review.
I've made these checks stricter and also clearer for users to follow.

syun64 commented on pyiceberg/io/pyarrow.py:
```diff
@@ -1296,31 +1297,49 @@ def to_requested_schema(
-                    and values.type.unit == "ns"
-                ):
-                    return values.cast(target_type, safe=False)
+                if field.field_type == TimestampType():
```
stricter and clearer `field_type` driven checks

syun64 commented on tests/io/test_pyarrow.py:
```diff
@@ -1798,3 +1799,35 @@ def test_identity_partition_on_multi_columns() -> None:
         ]) == arrow_table.sort_by([("born_year", "ascending"), ("n_legs", "ascending"), ("animal", "ascending")])
+
+
+def test_to_requested_schema_timestamps(
```
I noticed that there weren't any tests for `to_requested_schema` although it is a public API, so I added this in.

Do we want to keep `to_requested_schema` as a public API?

HonahX replied:
+1 on making `to_requested_schema` private. IMO this seem to be specific to our internal workflow so making it private can give us more flexibility in future development.

Actually, the signature of this API has been updated in https://github.com/apache/iceberg-python/pull/786: `pa.Table` to `pa.RecordBatch`, which is a breaking change. I think it may be a good time to send a deprecation notice on the old one and making the new one private. WDYT?

syun64 replied:
I've made the function private on this PR, and have another one up to port back to public API function with the previous signature with a deprecation notice here: https://github.com/apache/iceberg-python/pull/918

kevinjqliu reviewed (approved):
LGTM, glad we're getting rid of the dataframe casting. 

As a followup, let's investigate the level of abstraction on the write path. Currently we are doing schema compatible check, schema coersion, bin-packing, etc at different levels of the stack. It'll be good to optimize and see which functions can be pushed up the stack. 

For example, here's what the `overwrite` path looks like
```
overwrite
	_dataframe_to_data_files
		write_file
			write_parquet 
```

HonahX commented:
Merged! Thanks for the great work from @syun64 and reviews from @Fokko @kevinjqliu
